Tests/GRDBTests/MutablePersistableRecordEncodableTests.swift

```diff
@@ -800,5 +800,60 @@ extension MutablePersistableRecordEncodableTests {
                 "nestedUnkeyed": "[\"baz\",null,\"JSON nested: nestedUnkeyed\"]"])
         }
     }
+
+    func testUserInfoJsonEncoding() throws {
+        struct NestedStruct : Codable {
+            let firstName: String?
+            let lastName: String?
+
+            init(firstName: String?, lastName: String?) {
+                self.firstName = firstName
+                self.lastName = lastName
+            }
+
+            func encode(to encoder: Encoder) throws {
+                let userInfoValue = encoder.userInfo[.testKey] as? String
+                var container = encoder.container(keyedBy: CodingKeys.self)
+
+                if userInfoValue == "correct" {
+                    try container.encode(firstName, forKey: .firstName)
+                    try container.encode(lastName, forKey: .lastName)
+                } else {
+                    try container.encode(firstName, forKey: .lastName)
+                    try container.encode(lastName, forKey: .firstName)
+                }
+            }
+        }
+
+        struct StructWithNestedType : PersistableRecord, FetchableRecord, Codable {
+            static let databaseTableName = "t1"
+            static var databaseEncodingUserInfo: [CodingUserInfoKey: Any] = [CodingUserInfoKey.testKey: "correct"]
```

Comment:
Writing just `static var databaseEncodingUserInfo = [CodingUserInfoKey.testKey: "correct"]` will lead to unexpected behavior. The `databaseEncodingUserInfo` dictionary will still be empty at the time it's passed to the jsonEncoder at https://github.com/groue/GRDB.swift/blob/47199ed72b6230fb498897eaaf6b4e85d1d2f2c5/GRDB/Record/FetchableRecord.swift#L123

Reply:
Yes. Without the explicit type, the compiler sees `databaseEncodingUserInfo: [CodingUserInfoKey: String]`, which does not match the protocol requirement.